### PR TITLE
Metaprogramming

### PR DESCRIPTION
**Pull request by Vretamoza**

### CODE REVIEW

File: metaprogramming/quote_unquote.exs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# QUOTING
+IO.inspect(quote do: sum(1, 2, 3))
```
Comment (juanhurtado10) [inline metaprogramming/quote_unquote.exs:2]: @Vretamoza how do the quote and unquote works?

File: metaprogramming/domain_specific_languages.exs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  # This will be invoked right before the target module is compiled
+  # giving us the perfect opportunity to inject the `run/0` function
+  @doc false
+  defmacro __before_compile__(_env) do
+    quote do
+      def run do
+        Enum.each @tests, fn name ->
+          IO.puts "Running #{name}"
+          apply(__MODULE__, name, [])
```
Comment (juanhurtado10) [inline metaprogramming/domain_specific_languages.exs:42]: @Vretamoza what does this line of code do?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

**Comment by juanhurtado10** [comment]
@Vretamoza a rebase is required! Thanks

File: metaprogramming/quote_unquote.exs (hunk 1 above)
Comment (Vretamoza) [reply metaprogramming/quote_unquote.exs:2]: quotes returns the data structure representation of a function. With unquote we can inject code inside our quoted expressions.

File: metaprogramming/domain_specific_languages.exs (hunk 2 above)
Comment (Vretamoza) [reply metaprogramming/domain_specific_languages.exs:42]: It is a function that allows us to run all the defined tests.

Comment (juanhurtado10) [reply metaprogramming/domain_specific_languages.exs:42]: tests? are you sure?

**Review by juanhurtado10 — APPROVED** [review]

File: metaprogramming/domain_specific_languages.exs (hunk 2 above)
Comment (Vretamoza) [reply metaprogramming/domain_specific_languages.exs:42]: It saves the name of all defined tests in a module attribute so they can be executed by the run function. That's what I understood.

Comment (juanhurtado10) [reply metaprogramming/domain_specific_languages.exs:42]: @Vretamoza I was expecting an answer about the `apply/3` function which dynamically executes functions for the given module, but it is OK :)

**Pull request merged**